dynamic shapes builder API

avikchaudhuri opened the pull request:
This PR introduces a new way of building `dynamic_shapes` for export. The idea is to build up a mapping from input tensors to the dynamic shapes that should be assigned to their corresponding fake tensors. 

This mapping is automatically converted to the current form of `dynamic_shapes`, which must exactly match the structure of inputs. We do this by using pytree utils.

With the current `dynamic_shapes`, we had to be careful about user-defined classes that are registered with pytree, since  such classes are not necessarily polymorphic containers; they may be fine containing tensors, but not dynamic shapes. Thus we had decided to allow input instances of such classes to be associated with dynamic shapes in flattened form. This decision needs to be mirrored in this PR as well. To make it easier to keep these code paths in sync, we refactor the current recursive procedure for associating inputs with dynamic shapes to use the same pytree utils. This needs minor fixes to a few tests where `dynamic_shapes` were not exactly matching the structure of inputs.

Differential Revision: D56551992

JackCaoG commented:
Can you update the XLA pin in https://github.com/pytorch/pytorch/blob/main/.github/ci_commit_pins/xla.txt to `75278161c81f9308d412690b3e4d7cb23f35e8ab`? CI should start passing.

avikchaudhuri commented on .github/ci_commit_pins/xla.txt:
```diff
@@ -1 +1 @@
-58a412cb271a3f98ae2e01fd1d24bdbb66645d4e
+75278161c81f9308d412690b3e4d7cb23f35e8ab
```
See https://github.com/pytorch/pytorch/pull/124898#issuecomment-2083330717

avikchaudhuri committed:
dynamic shapes builder API

Summary:
This PR introduces a new way of building dynamic_shapes for export. The idea is to build up a mapping from input tensors to the dynamic shapes that should be assigned to their corresponding fake tensors.

This mapping is automatically converted to the current form of dynamic_shapes, which must exactly match the structure of inputs. We do this by using pytree utils.

With the current dynamic_shapes, we had to be careful about user-defined classes that are registered with pytree, since such classes are not necessarily polymorphic containers; they may be fine containing tensors, but not dynamic shapes. Thus we had decided to allow input instances of such classes to be associated with dynamic shapes in flattened form. This decision needs to be mirrored in this PR as well. To make it easier to keep these code paths in sync, we refactor the current recursive procedure for associating inputs with dynamic shapes to use the same pytree utils. This needs minor fixes to a few tests where dynamic_shapes were not exactly matching the structure of inputs.

Test Plan: added / updated tests

Reviewed By: zhxchen17

Differential Revision: D56551992
